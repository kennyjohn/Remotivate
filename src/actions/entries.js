import database from '../firebase/firebase';

// Action Creators
// ADD_ENTRY
export const addEntry = (entry) => ({
  type: 'ADD_ENTRY',
  entry
});

export const startAddEntry = (entryData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const { 
      text = '',
      author = '',
      video = '',
      image = '',
      purpose = '',
      createdAt = 0
    } = entryData;

    const entry = { text, author, video, image, purpose, createdAt };
    return database.ref(`users/${uid}/entries`).push(entry)
    .then((ref) => {
      dispatch(addEntry({
        id: ref.key,
        ...entry
      }));
    });
  };
};

// REMOVE_ENTRY
export const removeEntry = ({ id } = {}) => ({
    type: 'REMOVE_ENTRY',
    id
});

export const startRemoveEntry = ({ id } = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/entries/${id}`).remove()
        .then(() => dispatch(removeEntry({ id })));
    };
};

// EDIT_ENTRY
export const editEntry = (id, updates) => ({
    type: 'EDIT_ENTRY',
    id,
    updates
});

export const startEditEntry = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/entries/${id}`).update(updates)
        .then(() => dispatch(editEntry(id, updates)))
    };
}

export const setEntries = (entries) => ({
  type: 'SET_ENTRIES',
  entries
});

export const startSetEntries = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/entries`).once('value').then((snapshot) => {
      const entries = [];
      snapshot.forEach((childSnapshot) => {
        entries.push({
          id: childSnapshot.key,
          ...childSnapshot.val()
        });
      dispatch(setEntries(entries));
      });
    });
  };
};