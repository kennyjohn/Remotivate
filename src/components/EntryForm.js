import React from 'react';
import moment from 'moment';

export default class EntryForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            text: props.entry ? props.entry.text : '',
            author: props.entry ? props.entry.author : '',
            video: props.entry ? props.entry.video : '',
            image: props.entry ? props.entry.image : '',
            purpose: props.entry ? props.entry.purpose : '',
            createdAt: props.entry ? moment(props.entry.createdAt) : moment(),
            disabledTextFlag: true, 
            disabledImageFlag: true, 
            disabledVideoFlag: true,
            error: ''
        };
    }  
    onTextChange = (e) => {
        const text = e.target.value;
        this.setState(() => ({ text }));
    }
    onAuthorChange = (e) => {
        const author = e.target.value;
        this.setState(() => ({ author }));
    }
    onVideoURLChange = (e) => {
        const url = e.target.value;
        this.setState(() => ({ video: url }));
    }
    onImageURLChange = (e) => {
        const url = e.target.value;
        this.setState(() => ({ image: url }));
    }
    onPurposeChange = (e) => {
        const purpose = e.target.value;
        this.setState(() => ({ purpose }));
    }
    onSubmit = (e) => {
        e.preventDefault();
        const clicked = this.props.buttonClicked;
        if(!this.state.text && clicked === 'text') {
            this.setState(() => ({ error: 'Please provide text for your text entry.' }));
            return;
        }
        else if(!this.state.video && clicked === 'video') {
            this.setState(() => ({ error: 'Please provide a URL for your video entry.' }));
            return;
        }
        else if(!this.state.image && clicked === 'image') {
            this.setState(() => ({ error: 'Please provide a URL for your image entry.' }));
            return;
        }
        else {
            this.setState(() => ({ error: '' }));
            this.props.onSubmit({
                text: this.state.text,
                author: this.state.author,
                video: this.state.video,
                image: this.state.image,
                purpose: this.state.purpose,
                createdAt: this.state.createdAt.valueOf()
            });  
        }
    };
    componentDidMount() {
        if(this.state.text) {
            this.setState(() => ({
                disabledTextFlag: false,
                disabledImageFlag: true,
                disabledVideoFlag: true
            }));
        } else if(this.state.image) {
            this.setState(() => ({
                disabledTextFlag: true,
                disabledImageFlag: false,
                disabledVideoFlag: true
            }));
        } else {
            this.setState(() => ({
                disabledTextFlag: true,
                disabledImageFlag: true,
                disabledVideoFlag: false
            }));
        }
    }
    render() {
        return (
            <div>
                {this.props.buttonClicked && <h1 className="form__entry-header">Entry Form</h1>}
                <form className="form" onSubmit={this.onSubmit}>
                    {this.state.error && <p className="form__error">{this.state.error}</p>}
                    <textarea
                    /* disabled attribute accepts single attribute boolean (e.g. true, false);
                    can't refer to a function to return boolean result */
                    maxLength="1000"
                    disabled={!this.props.buttonClicked ? this.state.disabledTextFlag : this.props.buttonClicked !== "text"}
                    placeholder="Text"
                    onChange={this.onTextChange}
                    value={this.state.text} 
                    className="textarea"
                    >
                    </textarea>
                    <input 
                    maxLength="70"
                    type="text"
                    placeholder="Author (Optional)"
                    onChange={this.onAuthorChange}
                    value={this.state.author}
                    className="text-input"
                    />
                    <input
                    disabled={!this.props.buttonClicked ? this.state.disabledVideoFlag : this.props.buttonClicked !== "video"}
                    type="text"
                    placeholder="Youtube Video URL"
                    onChange={this.onVideoURLChange}
                    value={this.state.video}
                    className="text-input"
                    />
                    <input
                    disabled={!this.props.buttonClicked ? this.state.disabledImageFlag : this.props.buttonClicked !== "image"}
                    type="text"
                    placeholder="Image URL"
                    onChange={this.onImageURLChange}
                    value={this.state.image}
                    className="text-input"
                    />
                    <textarea
                    placeholder={`Why is this ${!this.state.disabledTextFlag || this.props.buttonClicked === 'text' ? 'text' : 
                    !this.state.disabledImageFlag || this.props.buttonClicked === 'image' ? 'image' :
                    !this.state.disabledVideoFlag || this.props.buttonClicked === 'video' ? 'video' :
                    ''} important to you?`}
                    onChange={this.onPurposeChange}
                    value={this.state.purpose}
                    className="textarea"
                    >
                    </textarea>
                    <div>
                        <button className="button button--save">Save Entry</button>
                    </div>
                </form>
            </div>
        );
    }
}