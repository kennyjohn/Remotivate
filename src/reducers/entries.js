const entriesReducerDefaultState = [];

export default (state = entriesReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_ENTRY':
            return [
                ...state,
                action.entry
            ];
        case 'REMOVE_ENTRY':
            return state.filter(({ id }) => id !== action.id);
        case 'EDIT_ENTRY':
            return state.map((entry) => {
                if(entry.id === action.id) {
                    return [
                        ...state,
                        ...action.updates
                    ];
                } else {
                    return entry;
                }
            });
        case 'SET_ENTRIES':
            return action.entries;
        default:
            return state;
    }
}