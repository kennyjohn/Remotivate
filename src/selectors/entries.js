import moment from 'moment';

export default (entries, { value, sortByDate, startDate, endDate }) => {
    return entries.filter((entry) => {
        const createdAtMoment = moment(entry.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
        const textFilterMatch = entry.text ? true : false;
        const videoFilterMatch = entry.video ? true : false;
        const imageFilterMatch = entry.image ? true : false;
        if(value === "text") {
            return startDateMatch && endDateMatch && textFilterMatch;
        } else if (value === "video") {
            return startDateMatch && endDateMatch && videoFilterMatch;
        } else if (value === "image") {
            return startDateMatch && endDateMatch && imageFilterMatch;
        } else {
            return startDateMatch && endDateMatch;
        }
    }).sort((a, b) => {
      if (sortByDate) {
        return a.createdAt < b.createdAt ? 1 : -1;
      }
    });
};
  