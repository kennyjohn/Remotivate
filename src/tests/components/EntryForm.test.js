import React from 'react';
import { shallow } from 'enzyme';
import entries from '../fixtures/entries';
import EntryForm from '../../components/EntryForm';

test('should render EntryForm', () => {
    const wrapper = shallow(<EntryForm></EntryForm>);
    expect(wrapper).toMatchSnapshot();
});

test('should render EntryForm with entry data', () => {
    const wrapper = shallow(<EntryForm entry={entries[0]}></EntryForm>);
    expect(wrapper).toMatchSnapshot();
});

test('should render error for invalid text submission', () => {
    const wrapper = shallow(<EntryForm buttonClicked="text"></EntryForm>);
    expect(wrapper).toMatchSnapshot();
    wrapper.find('form').simulate('submit', {
        preventDefault: () => { }
    });
    expect(wrapper.state('error')).toEqual('Please provide text for your text entry.');
    expect(wrapper).toMatchSnapshot();
});

test('should render error for invalid image submission', () => {
    const wrapper = shallow(<EntryForm buttonClicked="image"></EntryForm>);
    expect(wrapper).toMatchSnapshot();
    wrapper.find('form').simulate('submit', {
        preventDefault: () => { }
    });
    expect(wrapper.state('error')).toEqual('Please provide a URL for your image entry.');
    expect(wrapper).toMatchSnapshot();
});

test('should render error for invalid video submission', () => {
    const wrapper = shallow(<EntryForm buttonClicked="video"></EntryForm>);
    expect(wrapper).toMatchSnapshot();
    wrapper.find('form').simulate('submit', {
        preventDefault: () => { }
    });
    expect(wrapper.state('error')).toEqual('Please provide a URL for your video entry.');
    expect(wrapper).toMatchSnapshot();
});

test('should set text on input change', () => {
    const value = 'New text';
    const wrapper = shallow(<EntryForm></EntryForm>);
    wrapper.find('textarea').at(0).simulate('change', {
        target: { value }
    });
    expect(wrapper.state('text')).toBe(value);
});

test('should set purpose on input change', () => {
    const value = 'New purpose';
    const wrapper = shallow(<EntryForm></EntryForm>);
    wrapper.find('textarea').at(1).simulate('change', {
        target: { value }
    });
    expect(wrapper.state('purpose')).toBe(value);
});

test('should set author on input change', () => {
    const value = 'Jay Shetty';
    const wrapper = shallow(<EntryForm></EntryForm>);
    wrapper.find('input').at(0).simulate('change', {
        target: { value }
    });
    expect(wrapper.state('author')).toBe(value);
});

test('should set video URL on input change', () => {
    const value = 'video.url.com';
    const wrapper = shallow(<EntryForm></EntryForm>);
    wrapper.find('input').at(1).simulate('change', {
        target: { value }
    });
    expect(wrapper.state('video')).toBe(value);
});

test('should set image URL on input change', () => {
    const value = 'image.url.com';
    const wrapper = shallow(<EntryForm></EntryForm>);
    wrapper.find('input').at(2).simulate('change', {
        target: { value }
    });
    expect(wrapper.state('image')).toBe(value);
});

test('should call onSubmit prop for valid video submission', () => {
    const onSubmitSpy = jest.fn();
    const wrapper = shallow(<EntryForm
        entry={entries[2]}
        onSubmit={onSubmitSpy}
        ></EntryForm>);
    wrapper.find('form').simulate('submit', {
        preventDefault: () => {}
    });
    expect(wrapper.state('error')).toBe('');
    expect(onSubmitSpy).toHaveBeenLastCalledWith({
        text: entries[2].text,
        author: entries[2].author,
        video: entries[2].video,
        image: entries[2].image,
        createdAt: entries[2].createdAt,
        purpose: entries[2].purpose
    });
});

test('should call onSubmit prop for valid text submission', () => {
    const onSubmitSpy = jest.fn();
    const wrapper = shallow(<EntryForm
        entry={entries[0]}
        onSubmit={onSubmitSpy}
        ></EntryForm>);
    wrapper.find('form').simulate('submit', {
        preventDefault: () => {}
    });
    expect(wrapper.state('error')).toBe('');
    expect(onSubmitSpy).toHaveBeenLastCalledWith({
        text: entries[0].text,
        author: entries[0].author,
        video: entries[0].video,
        image: entries[0].image,
        createdAt: entries[0].createdAt,
        purpose: entries[0].purpose
    });
});

test('should call onSubmit prop for valid image submission', () => {
    const onSubmitSpy = jest.fn();
    const wrapper = shallow(<EntryForm
        entry={entries[1]}
        onSubmit={onSubmitSpy}
        ></EntryForm>);
    wrapper.find('form').simulate('submit', {
        preventDefault: () => {}
    });
    expect(wrapper.state('error')).toBe('');
    expect(onSubmitSpy).toHaveBeenLastCalledWith({
        text: entries[1].text,
        author: entries[1].author,
        video: entries[1].video,
        image: entries[1].image,
        createdAt: entries[1].createdAt,
        purpose: entries[1].purpose
    });
});