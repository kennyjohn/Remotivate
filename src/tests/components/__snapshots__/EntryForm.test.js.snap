// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EntryForm 1`] = `
<div>
  <h1
    className="form__entry-header"
  >
    Entry Form
  </h1>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <textarea
      className="textarea"
      disabled={true}
      maxLength="1000"
      onChange={[Function]}
      placeholder="Text"
      value=""
    />
    <input
      className="text-input"
      maxLength="70"
      onChange={[Function]}
      placeholder="Author (Optional)"
      type="text"
      value=""
    />
    <input
      className="text-input"
      disabled={true}
      onChange={[Function]}
      placeholder="Youtube Video URL"
      type="text"
      value=""
    />
    <input
      className="text-input"
      disabled={true}
      onChange={[Function]}
      placeholder="Image URL"
      type="text"
      value=""
    />
    <textarea
      className="textarea"
      onChange={[Function]}
      placeholder="Why is this  important to you?"
      value=""
    />
    <div>
      <button
        className="button button--save"
      >
        Save Entry
      </button>
    </div>
  </form>
</div>
`;

exports[`should render EntryForm with entry data 1`] = `
<div>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <textarea
      className="textarea"
      disabled={false}
      maxLength="1000"
      onChange={[Function]}
      placeholder="Text"
      value="testing"
    />
    <input
      className="text-input"
      maxLength="70"
      onChange={[Function]}
      placeholder="Author (Optional)"
      type="text"
      value="Kobe Bryant"
    />
    <input
      className="text-input"
      disabled={true}
      onChange={[Function]}
      placeholder="Youtube Video URL"
      type="text"
      value=""
    />
    <input
      className="text-input"
      disabled={true}
      onChange={[Function]}
      placeholder="Image URL"
      type="text"
      value=""
    />
    <textarea
      className="textarea"
      onChange={[Function]}
      placeholder="Why is this text important to you?"
      value=""
    />
    <div>
      <button
        className="button button--save"
      >
        Save Entry
      </button>
    </div>
  </form>
</div>
`;

exports[`should render error for invalid image submission 1`] = `
<div>
  <h1
    className="form__entry-header"
  >
    Entry Form
  </h1>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <textarea
      className="textarea"
      disabled={true}
      maxLength="1000"
      onChange={[Function]}
      placeholder="Text"
      value=""
    />
    <input
      className="text-input"
      maxLength="70"
      onChange={[Function]}
      placeholder="Author (Optional)"
      type="text"
      value=""
    />
    <input
      className="text-input"
      disabled={true}
      onChange={[Function]}
      placeholder="Youtube Video URL"
      type="text"
      value=""
    />
    <input
      className="text-input"
      disabled={false}
      onChange={[Function]}
      placeholder="Image URL"
      type="text"
      value=""
    />
    <textarea
      className="textarea"
      onChange={[Function]}
      placeholder="Why is this image important to you?"
      value=""
    />
    <div>
      <button
        className="button button--save"
      >
        Save Entry
      </button>
    </div>
  </form>
</div>
`;

exports[`should render error for invalid image submission 2`] = `
<div>
  <h1
    className="form__entry-header"
  >
    Entry Form
  </h1>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <p
      className="form__error"
    >
      Please provide a URL for your image entry.
    </p>
    <textarea
      className="textarea"
      disabled={true}
      maxLength="1000"
      onChange={[Function]}
      placeholder="Text"
      value=""
    />
    <input
      className="text-input"
      maxLength="70"
      onChange={[Function]}
      placeholder="Author (Optional)"
      type="text"
      value=""
    />
    <input
      className="text-input"
      disabled={true}
      onChange={[Function]}
      placeholder="Youtube Video URL"
      type="text"
      value=""
    />
    <input
      className="text-input"
      disabled={false}
      onChange={[Function]}
      placeholder="Image URL"
      type="text"
      value=""
    />
    <textarea
      className="textarea"
      onChange={[Function]}
      placeholder="Why is this image important to you?"
      value=""
    />
    <div>
      <button
        className="button button--save"
      >
        Save Entry
      </button>
    </div>
  </form>
</div>
`;

exports[`should render error for invalid text submission 1`] = `
<div>
  <h1
    className="form__entry-header"
  >
    Entry Form
  </h1>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <textarea
      className="textarea"
      disabled={false}
      maxLength="1000"
      onChange={[Function]}
      placeholder="Text"
      value=""
    />
    <input
      className="text-input"
      maxLength="70"
      onChange={[Function]}
      placeholder="Author (Optional)"
      type="text"
      value=""
    />
    <input
      className="text-input"
      disabled={true}
      onChange={[Function]}
      placeholder="Youtube Video URL"
      type="text"
      value=""
    />
    <input
      className="text-input"
      disabled={true}
      onChange={[Function]}
      placeholder="Image URL"
      type="text"
      value=""
    />
    <textarea
      className="textarea"
      onChange={[Function]}
      placeholder="Why is this text important to you?"
      value=""
    />
    <div>
      <button
        className="button button--save"
      >
        Save Entry
      </button>
    </div>
  </form>
</div>
`;

exports[`should render error for invalid text submission 2`] = `
<div>
  <h1
    className="form__entry-header"
  >
    Entry Form
  </h1>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <p
      className="form__error"
    >
      Please provide text for your text entry.
    </p>
    <textarea
      className="textarea"
      disabled={false}
      maxLength="1000"
      onChange={[Function]}
      placeholder="Text"
      value=""
    />
    <input
      className="text-input"
      maxLength="70"
      onChange={[Function]}
      placeholder="Author (Optional)"
      type="text"
      value=""
    />
    <input
      className="text-input"
      disabled={true}
      onChange={[Function]}
      placeholder="Youtube Video URL"
      type="text"
      value=""
    />
    <input
      className="text-input"
      disabled={true}
      onChange={[Function]}
      placeholder="Image URL"
      type="text"
      value=""
    />
    <textarea
      className="textarea"
      onChange={[Function]}
      placeholder="Why is this text important to you?"
      value=""
    />
    <div>
      <button
        className="button button--save"
      >
        Save Entry
      </button>
    </div>
  </form>
</div>
`;

exports[`should render error for invalid video submission 1`] = `
<div>
  <h1
    className="form__entry-header"
  >
    Entry Form
  </h1>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <textarea
      className="textarea"
      disabled={true}
      maxLength="1000"
      onChange={[Function]}
      placeholder="Text"
      value=""
    />
    <input
      className="text-input"
      maxLength="70"
      onChange={[Function]}
      placeholder="Author (Optional)"
      type="text"
      value=""
    />
    <input
      className="text-input"
      disabled={false}
      onChange={[Function]}
      placeholder="Youtube Video URL"
      type="text"
      value=""
    />
    <input
      className="text-input"
      disabled={true}
      onChange={[Function]}
      placeholder="Image URL"
      type="text"
      value=""
    />
    <textarea
      className="textarea"
      onChange={[Function]}
      placeholder="Why is this video important to you?"
      value=""
    />
    <div>
      <button
        className="button button--save"
      >
        Save Entry
      </button>
    </div>
  </form>
</div>
`;

exports[`should render error for invalid video submission 2`] = `
<div>
  <h1
    className="form__entry-header"
  >
    Entry Form
  </h1>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <p
      className="form__error"
    >
      Please provide a URL for your video entry.
    </p>
    <textarea
      className="textarea"
      disabled={true}
      maxLength="1000"
      onChange={[Function]}
      placeholder="Text"
      value=""
    />
    <input
      className="text-input"
      maxLength="70"
      onChange={[Function]}
      placeholder="Author (Optional)"
      type="text"
      value=""
    />
    <input
      className="text-input"
      disabled={false}
      onChange={[Function]}
      placeholder="Youtube Video URL"
      type="text"
      value=""
    />
    <input
      className="text-input"
      disabled={true}
      onChange={[Function]}
      placeholder="Image URL"
      type="text"
      value=""
    />
    <textarea
      className="textarea"
      onChange={[Function]}
      placeholder="Why is this video important to you?"
      value=""
    />
    <div>
      <button
        className="button button--save"
      >
        Save Entry
      </button>
    </div>
  </form>
</div>
`;
